compiler/resolution/implementForallIntents.cpp

```diff
@@ -1778,38 +1878,221 @@ static bool isFsIndexVar(ForallStmt* fs, Symbol* sym)
 // A forall-intents variation on findOuterVars() in createTaskFunctions.cpp:
 // Find all symbols used in 'block' and defined outside of it.
 //
-static void findOuterVarsNew(ForallStmt* fs, BlockStmt* block, SymbolMap& uses) {
+static void findOuterVarsNew(ForallStmt* fs, BlockStmt* block,
+                             FIIvec& fivec, FIImap& fimap)
+{
   std::vector<SymExpr*> symExprs;
 
   collectSymExprs(block, symExprs);
 
   for_vector(SymExpr, symExpr, symExprs) {
     Symbol* sym = symExpr->symbol();
 
-    if (isLcnSymbol(sym))
-      if (!isFsIndexVar(fs, sym) && isOuterVarNew(sym, block))
-        uses.put(sym, markUnspecified);
+    if (isLcnSymbol(sym) &&
+        // exclude a few things
+        !isFsIndexVar(fs, sym) &&
+        isOuterVarNew(sym, block) &&
+        sym->type != dtMethodToken &&
+        !sym->hasFlag(FLAG_ARG_THIS)
+       )
+      // OK, add it
+      addFII(sym, fivec, fimap);
+  }
+}
+
+/////////////////////////////////////////////////////////////////////////////
+// markOuterVarsWithIntentsNew() and helpers
+//
+
+// Finalize the reduction:  outerVar = globalOp.generate()
+static void insertGenerateAfter(Expr* ref, Symbol* fiVarSym, Symbol* globalOp) {
+  Expr* next = ref->next; // nicer ordering of the following insertions
+  INT_ASSERT(next);
+  VarSymbol* genTemp = newTemp("chpl_gentemp");
+  next->insertBefore(new DefExpr(genTemp));
+  next->insertBefore("'move'(%S, generate(%S,%S))",
+                     genTemp, gMethodToken, globalOp);
+  next->insertBefore(new CallExpr("=", fiVarSym, genTemp));
+}
+
+static Symbol* setupRiGlobalOp(ForallStmt* fs, Symbol* fiVarSym,
+                               Expr* origRiSpec, TypeSymbol* riTypeSym,
+                               Expr* eltTypeArg)
+{
+  BlockStmt* hld = new BlockStmt(); // "holder"
+  fs->insertBefore(hld);
+      
+  VarSymbol* globalOp =  newTemp("chpl__reduceGlob");
+  hld->insertAtTail(new DefExpr(globalOp));
+
+  if (!eltTypeArg) {
+    Type* eltType   = fiVarSym->qualType().type()->getValType();
+    // fiVar should have been resolved already
+    INT_ASSERT(eltType != dtUnknown && eltType != dtAny);
+
+    if (eltType->symbol->hasFlag(FLAG_HAS_RUNTIME_TYPE)) {
+      // arg = typeof(fiVar)
+      VarSymbol* typeofTemp = newTemp("chpl__typeofET");
+      typeofTemp->addFlag(FLAG_TYPE_VARIABLE);
+      hld->insertAtTail(new DefExpr(typeofTemp));
+      hld->insertAtTail("'move'(%S,'typeof'(%S))", typeofTemp, fiVarSym);
+      eltTypeArg = new SymExpr(typeofTemp);
+    } else {
+      // Forego typeof(), use the type directly.
+      eltTypeArg = new SymExpr(eltType->symbol);
+    }
+  }
```

Comment:
I can see that the above two were moved, but it's hard for me to tell if you also changed them. Such rearrangements are best done in separate commits so they don't distract from the other changes. For now, just let me know if you changed them.